x-pack/filebeat/input/cometd/input.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/common/transport/httpcommon"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+const (
+	inputName = "cometd"
+)
+
+type cometdInput struct {
+	config
+
+	log      *logp.Logger
+	inputCtx context.Context // Wraps the Done channel from parent input.Context.
+
+	workerCtx    context.Context    // Worker goroutine context. It's cancelled when the input stops or the worker exits.
+	workerCancel context.CancelFunc // Used to signal that the worker should stop.
+	workerOnce   sync.Once          // Guarantees that the worker goroutine is only started once.
+	workerWg     sync.WaitGroup     // Waits on pubsub worker goroutine.
+
+	ackedCount *atomic.Uint32                   // Total number of successfully ACKed pubsub messages.
+	Transport  httpcommon.HTTPTransportSettings `config:",inline"`
+	Retry      retryConfig                      `config:"retry"`
+}
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(errors.Wrapf(err, "failed to register %v input", inputName))
+	}
+}
+
+// NewInput creates a new CometD Pub/Sub input that consumes events from
+// a topic subscription.
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (inp input.Input, err error) {
+	// Extract and validate the input's configuration.
+	conf := defaultConfig()
+	if err = cfg.Unpack(&conf); err != nil {
+		return nil, fmt.Errorf("cometd config: unpacking of config failed: %v", err)
+	}
+
+	logger := logp.NewLogger("cometd").With(
+		"pubsub_channel", conf.ChannelName)
+
+	// Wrap input.Context's Done channel with a context.Context. This goroutine
+	// stops with the parent closes the Done channel.
+	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+	go func() {
+		defer cancelInputCtx()
+		select {
+		case <-inputContext.Done:
+		case <-inputCtx.Done():
+		}
+	}()
+
+	// If the input ever needs to be made restartable, then context would need
+	// to be recreated with each restart.
+	workerCtx, workerCancel := context.WithCancel(inputCtx)
+
+	in := &cometdInput{
+		config:       conf,
+		log:          logger,
+		inputCtx:     inputCtx,
+		workerCtx:    workerCtx,
+		workerCancel: workerCancel,
+		ackedCount:   atomic.NewUint32(0),
+	}
+
+	in.log.Info("Initialized cometD input.")
+	return in, nil
+}
+
+// Run starts the pubsub input worker then returns. Only the first invocation
+// will ever start the pubsub worker.
+func (in *cometdInput) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.log.Info("Pub/Sub input worker has started.")
+			defer in.log.Info("Pub/Sub input worker has stopped.")
+			defer in.workerWg.Done()
+			defer in.workerCancel()
+			if err := in.run(); err != nil {
+				in.log.Error(err)
+				return
+			}
+		}()
+	})
+}
+
+// Stop stops the pubsub input and waits for it to fully stop.
```

Comment:
AFAIR It should be the other way round: `Stop` just stops the logic, but doesn't wait, but `Wait` can call the `Stop` and wait until all processing is done.